File: self-serve-storage/python/daaas_storage.py
#!/usr/bin/env python3

#########################################
###                                   ###
###          daaas_storage.py         ###
###          ~~~~~~~~~~~~~~~~         ###
###                                   ###
###   Import this from your notebook  ###
###   to easily access your storage   ###
###                                   ###
#########################################

####################
###      API     ###
####################
###
###  instances =  get_instances()      # return all available MinIO instances
###  minio_client =  get_minio_client(instance)      # return a Minio object, pass in a valid MinIO instance
###
###  See: https://github.com/minio/minio-py

###########################
###    Usage Example    ###
###########################
###
###    import daaas_storage.py
###
###    instances = get_instances()
###    minio_client = get_minio_client(instances[0])    # pick an instance
###
###
###    # This minio client is from minio-py, and you can use it
###    # for and s3 purpose
###    # See https://github.com/minio/minio-py
###
###
###    # Example:
###    objects = minio_client.list_objects(
###        "shared",
###        prefix='blair-drummond/',
###        recursive=True
###          )
###
###    for obj in objects:
###        print(obj.bucket_name, obj.object_name.encode('utf-8'), obj.last_modified,
###              obj.etag, obj.size, obj.content_type)
###
###

import subprocess
from minio import Minio
import os
import json

def get_client(instance):
    """ Get the variables out of vault to create Minio Client. """

    with open(f'/vault/secrets/{instance}.json') as f:
        creds = json.load(f)
        minio_url = creds['MINIO_URL']

    import re
    # Get rid of http:// in minio URL
    http = re.compile('^https?://')

    # Create the minio client.
    s3Client = Minio(
        http.sub("", creds['MINIO_URL']),
        access_key=creds['MINIO_ACCESS_KEY'],
        secret_key=creds['MINIO_SECRET_KEY'],
        secure=minio_url.startswith('https'),
        region="us-west-1"
    )

    return s3Client

class Instances:
    instances = []
    def add_instance(self, instance):
        self.instances.append(instance)
        setattr(self, instance, get_client(instance)) 
    
    def __init__(self):
        instances = get_instances()
        for i in instances:
            self.add_instance(i)
            
def get_instances():
    files = os.listdir('/vault/secrets')
    instances = [file for file in files if not file.endswith('.json')]
    return instances

# instances = Instances() 

File: self-serve-storage/R/daaas_storage.R
#########################################
###                                   ###
###          daaas_storage.R          ###
###          ~~~~~~~~~~~~~~~          ###
###                                   ###
###   Source this from your notebook  ###
###   to easily access your storage   ###
###                                   ###
#########################################

###########################
###         API         ###
###########################
###
###   daaas_storage.get_client()      returns NULL
###   daaas_storage.get_instances()      returns NULL
###
###

###########################
###    Usage Example    ###
###########################
###
###    source("daaas_storage.R")
###
###    # Choose from
###
###    daaas_storage.get_client(instance) 
###    daaas_storage.get_instances()
###
###    ### Note, unlike the python version, these functions
###    ### modify the global option, instead of returning a
###    ### connection.
###
###    # Example:
###    get_bucket(bucket = "shared", use_https=FALSE, region="")
###

# Source the s3 storage secrets and urls.
get_bash_variable <- function (location, var) {
    system(
        sprintf(
            "bash -c 'source %s; echo $%s'",
            location,
            var
        ),
        intern = TRUE
    )
}
# Just sets the environment variables.
daaas_storage.get_client<- function(instance) {
    env_vars <- c("MINIO_URL", "MINIO_ACCESS_KEY", "MINIO_SECRET_KEY")
    location <- sprintf("/vault/secrets/%s", instance)
    minio <- if (requireNamespace("jsonlite", quietly = TRUE)) {
        # jsonlite is installed on the AAW's R image
        # works just as well with RJSONIO::fromJSON or rjson::fromJSON
        jsonlite::fromJSON(paste0(location, ".json"))
    } else {
        lapply(setNames(nm = env_vars), function(x) {
            system(sprintf("bash -c 'source %s; echo $%s'", location, x), intern = TRUE)
        })
    }
    Sys.setenv(
        "AWS_S3_ENDPOINT" = gsub("https?://", "", minio$MINIO_URL),
        "AWS_ACCESS_KEY_ID" = minio$MINIO_ACCESS_KEY,
        "AWS_SECRET_ACCESS_KEY" = minio$MINIO_SECRET_KEY,
        "AWS_DEFAULT_REGION" = "",
        "SECURE" = startsWith(minio$MINIO_URL, "https")
    )
}

# List MinIO instances
daaas_storage.get_instances <- function () {
    list <- grep(".*(?<!\\.json)$", list.files("/vault/secrets/"), perl=TRUE, value=TRUE)
    for (i in list) {
        print(i)
    }
}

